pkg/util/rbac/rbac.go

```diff
@@ -68,32 +68,90 @@ func ResourceRoleAssignmentWithName(roleID, spID, resourceType, resourceName, na
 }
 
-// ResourceGroupRoleAssignmentWithName returns a Resource granting roleID on the
-// current resource group to spID.  Arguments spID and name must be valid ARM
-// expressions, e.g. "'foo'" or "concat('foo')".  Use this function in ARM
-// templates which have already been deployed, to preserve the name and avoid a
-// RoleAssignmentExists error.
-func ResourceGroupRoleAssignmentWithName(roleID, spID string, name string, condition ...interface{}) *arm.Resource {
+func resourceGroupRoleAssignmentWithDetails(roleID, spID string, name string, dependsOn []string, subscriptionScope bool, condition ...interface{}) *arm.Resource {
+	var resourceIdFunction string
+	if subscriptionScope {
+		resourceIdFunction = "subscriptionResourceId"
+	} else {
+		resourceIdFunction = "resourceId"
+	}
 	r := &arm.Resource{
 		Resource: mgmtauthorization.RoleAssignment{
 			Name: to.StringPtr("[" + name + "]"),
 			Type: to.StringPtr("Microsoft.Authorization/roleAssignments"),
 			RoleAssignmentPropertiesWithScope: &mgmtauthorization.RoleAssignmentPropertiesWithScope{
 				Scope:            to.StringPtr("[resourceGroup().id]"),
-				RoleDefinitionID: to.StringPtr("[subscriptionResourceId('Microsoft.Authorization/roleDefinitions', '" + roleID + "')]"),
+				RoleDefinitionID: to.StringPtr("[" + resourceIdFunction + "('Microsoft.Authorization/roleDefinitions', '" + roleID + "')]"),
 				PrincipalID:      to.StringPtr("[" + spID + "]"),
 				PrincipalType:    mgmtauthorization.ServicePrincipal,
 			},
 		},
 		APIVersion: azureclient.APIVersion("Microsoft.Authorization"),
 	}
-
+	if dependsOn != nil {
+		r.DependsOn = dependsOn
+	}
 	if len(condition) > 0 {
 		r.Condition = condition[0]
 	}
+	return r
+}
+
+// ResourceGroupRoleAssignmentWithName returns a Resource granting roleID on the
+// current resource group to spID.  Arguments spID and name must be valid ARM
+// expressions, e.g. "'foo'" or "concat('foo')".  Use this function in ARM
+// templates which have already been deployed, to preserve the name and avoid a
+// RoleAssignmentExists error.
+func ResourceGroupRoleAssignmentWithName(roleID, spID string, name string, condition ...interface{}) *arm.Resource {
+	return resourceGroupRoleAssignmentWithDetails(roleID, spID, name, nil, true, condition...)
+}
+
+// ResourceGroupCustomRoleAssignmentWithName returns a Resource granting roleID on the
+// current resource group to spID.  Arguments spID and name must be valid ARM
+// expressions, e.g. "'foo'" or "concat('foo')".  Use this function in ARM
+// templates which have already been deployed, to preserve the name and avoid a
+// RoleAssignmentExists error. Additionally a dependency is added on the role, use this function
+// when the Role Definition is declared in the same ARM template at non-subscription scope.
+func ResourceGroupCustomRoleAssignmentWithName(roleID, spID string, name string, condition ...interface{}) *arm.Resource {
```

Comment:
there are no legacy cases to support here, so this function should not end in WithName and should use guid() to create a name.  Is that possible?